src/transaction/lock.rs

```diff
@@ -21,81 +21,92 @@ pub async fn resolve_locks(
+            let is_small_txn = lock.txn_size < 16;
+
+            let resolve_info = grouped
+                .entry((region.ver_id(), lock.lock_version))
+                .or_insert(ResolveInfo {
+                    region,
+                    primary_key: primary_key.clone(),
+                    keys: vec![],
+                });
+
+            if is_small_txn {
+                resolve_info.keys.push(primary_key);
+            } else {
+                //if txn is large,keep resolve_info.keys empty to send a full resolve lock request
+            }
+        } else {
             has_live_locks = true;
         }
-        expired
-    });
-
-    // records the commit version of each primary lock (representing the status of the transaction)
-    let mut commit_versions: HashMap<u64, u64> = HashMap::new();
-    let mut clean_regions: HashMap<u64, HashSet<RegionVerId>> = HashMap::new();
-    for lock in expired_locks {
-        let primary_key: Key = lock.primary_lock.into();
-        let region_ver_id = pd_client.region_for_key(&primary_key).await?.ver_id();
-        // skip if the region is cleaned
-        if clean_regions
-            .get(&lock.lock_version)
-            .map(|regions| regions.contains(&region_ver_id))
-            .unwrap_or(false)
-        {
-            continue;
-        }
+    }
 
-        let commit_version = match commit_versions.get(&lock.lock_version) {
-            Some(&commit_version) => commit_version,
-            None => {
-                let commit_version = requests::new_cleanup_request(primary_key, lock.lock_version)
-                    .execute(pd_client.clone())
-                    .await?;
-                commit_versions.insert(lock.lock_version, commit_version);
-                commit_version
-            }
-        };
+    for ((_region_ver_id, lock_version), resolve_info) in grouped {
+        let commit_version = requests::new_cleanup_request(resolve_info.primary_key, lock_version)
+            .execute(pd_client.clone())
+            .await?;
 
-        let cleaned_region = resolve_lock_with_retry(
-            lock.key.into(),
-            lock.lock_version,
+        resolve_lock_with_retry(
+            &resolve_info.region,
+            &resolve_info.keys,
+            lock_version,
             commit_version,
             pd_client.clone(),
         )
         .await?;
-        clean_regions
-            .entry(lock.lock_version)
-            .or_insert_with(HashSet::new)
-            .insert(cleaned_region);
     }
+
     Ok(!has_live_locks)
 }
 
 async fn resolve_lock_with_retry(
-    key: Key,
+    region: &Region,
+    keys: &Vec<Key>,
     start_version: u64,
     commit_version: u64,
     pd_client: Arc<impl PdClient>,
-) -> Result<RegionVerId> {
-    // TODO: Add backoff
+) -> Result<()> {
     let mut error = None;
+    let mut back_off = NoJitterBackoff::new(10, 1000, RESOLVE_LOCK_RETRY_LIMIT as u32);
+
     for _ in 0..RESOLVE_LOCK_RETRY_LIMIT {
```

Comment:
WIth the backoff providing retry limit (`next_delay_duration` returns `None` if the limit is exceeded), I think maybe we can use `loop` and `break` to avoid repeated reference to RESOLVE_LOCK_RETRY_LIMIT